File: 2021/day05.go
package main

import (
	"aoc/utils"
	"fmt"
	"strings"
)

type point struct{ x, y int }
type segment struct{ from, to point }

func visitedPoints(segments []segment, canVisit func(segment) bool) map[point]int {
	points := make(map[point]int)
	for _, segment := range segments {
		if canVisit(segment) {
			from, to := segment.from, segment.to
			dx, dy := utils.Sign(to.x-from.x), utils.Sign(to.y-from.y)
			for x, y := from.x, from.y; x != to.x || y != to.y; x, y = x+dx, y+dy {
				points[point{x, y}]++
			}
			points[point{to.x, to.y}]++
		}
	}
	return points
}

func countMultipleOverlaps(segments []segment, canVisit func(segment) bool) (c int) {
	for _, overlaps := range visitedPoints(segments, canVisit) {
		if overlaps > 1 {
			c++
		}
	}
	return
}

// Part1 of day5
func Part1(segments []segment) {
	fmt.Println(countMultipleOverlaps(segments, func(s segment) bool {
		return s.from.x == s.to.x || s.from.y == s.to.y
	}))
}

// Part2 of day5
func Part2(segments []segment) {
	fmt.Println(countMultipleOverlaps(segments, func(s segment) bool {
		return true
	}))
}

func main() {
	segments := []segment{}
	for _, line := range utils.ReadFileLinesOrDie("./input/day05.txt") {
		p := strings.Split(line, " -> ")
		c1, c2 := strings.Split(p[0], ","), strings.Split(p[1], ",")
		segments = append(segments, segment{
			from: point{
				x: utils.ToIntOrDie(c1[0]),
				y: utils.ToIntOrDie(c1[1]),
			},
			to: point{
				x: utils.ToIntOrDie(c2[0]),
				y: utils.ToIntOrDie(c2[1]),
			},
		})
	}
	Part1(segments)
	Part2(segments)
}

File: 2021/day05.py
def visit_points(points, from_coord, to_coord, with_dx, with_dy):
    x, y, x_end, y_end = *from_coord, *to_coord
    points[(x, y)] = 1 + points.get((x, y), 0)
    while x != x_end or y != y_end:
        x += with_dx
        y += with_dy
        points[(x, y)] = 1 + points.get((x, y), 0)


def sign(x):
    return 1 if x >= 0 else -1


def part1(coords):
    points = {}
    for coord in coords:
        x1, y1, x2, y2 = *coord[0], *coord[1]
        if x1 == x2:
            visit_points(points, from_coord=coord[0], to_coord=coord[1], with_dx=0, with_dy=sign(y2 - y1))
        elif y1 == y2:
            visit_points(points, from_coord=coord[0], to_coord=coord[1], with_dx=sign(x2 - x1), with_dy=0)

    print(sum(points[k] > 1 for k in points))


def part2(coords):
    points = {}
    for coord in coords:
        x1, y1, x2, y2 = *coord[0], *coord[1]
        if x1 == x2:
            visit_points(points, from_coord=coord[0], to_coord=coord[1], with_dx=0, with_dy=sign(y2 - y1))
        elif y1 == y2:
            visit_points(points, from_coord=coord[0], to_coord=coord[1], with_dx=sign(x2 - x1), with_dy=0)
        else:
            visit_points(points, from_coord=coord[0], to_coord=coord[1], with_dx=sign(x2 - x1), with_dy=sign(y2 - y1))

    print(sum(points[k] > 1 for k in points))


with open("input/day05.txt") as f:
    v = list(map(lambda l: list(map(lambda p: tuple(map(int, p.split(','))), l.split(' -> '))), f.readlines()))
    part1(v)
    part2(v)
